Launch SecretMq and encrypted query for tee contracts.

kvinwang opened the pull request:
The functions in pRuntime works now. See [Here](https://github.com/Phala-Network/phala-blockchain/commit/secret-mq-query-usage~) and [Here](https://github.com/Phala-Network/phala-blockchain/commit/secret-mq-query-usage) for some usage examples.

h4x3rotab commented on common/types/src/contract.rs:
```diff
@@ -0,0 +1,84 @@
+}
+
+pub fn command_topic(id: ContractId) -> Vec<u8> {
+    format!("phala/contract/{}/command", hex::encode(&id)).as_bytes().to_vec()
```
Do we assume commands are encrypted, while events are not?

kvinwang replied:
> Do we assume commands are encrypted, while events are not?

Exactly

h4x3rotab commented on common/types/src/lib.rs:
```diff
@@ -68,19 +71,27 @@ pub mod messaging {
         TransferToChain { dest: AccountId, value: Balance },
     }
 
+    impl<AccountId, Balance> BalanceCommand<AccountId, Balance> {
```
Oh, I think we should call it `Balances` to be aligned with Substrate's Balances pallet

h4x3rotab replied:
If we assume AccountId = sp_runtime::AccountId32, and Balance = u128, do you think it can simplify our code?

kvinwang replied:
> If we assume AccountId = sp_runtime::AccountId32, and Balance = u128, do you think it can simplify our code?

If we do so, such types will not be convenient to be used in pallets. Remember the `trait ToH256`?

kvinwang replied:
> Oh, I think we should call it `Balances` to be aligned with Substrate's Balances pallet

Done

h4x3rotab commented on standalone/pruntime/enclave/src/prpc_service.rs:
```diff
@@ -379,32 +385,43 @@ pub fn init_runtime(
-                let evt_mq = PeelingReceiver::new_plain(recv_mq.subscribe_bound());
+                // TODO.kevin: use real contract key
+                let contract_key = local_ecdh_key.clone();
+                let cmd_mq = PeelingReceiver::new_secret(
```
If we have two queues, is the order of the inputs preserved?

i.e. If on the blockchain we have [cmd1, ev1, cmd2], do we process them in order?

kvinwang replied:
> If we have two queues, is the order of the inputs preserved?

Yes, it is.

h4x3rotab replied:
Okay. That well fits our design

h4x3rotab reviewed (commented):
Found a problem while reviewing. Though our user (browser or client sdk), as well as the entities in pruntime (GK, worker, contracts) can send encrypted message, some cannot, like pallets, or the message from other blockchains. I saw you have separated the input to contracts into `command` and `event`, but I'm not sure if it's flexible enough.

Imagine a scenario. We create a private token contract to hold wallet balances. Usually we expect end user to send encrypted transfer commands to the contract. However, as a parachain, there could also be some other parachains that open accounts on our side, and manage the balances. They may not have the ability to send encrypted message, because they are just pallets and cannot deal with secrets. But even without privacy, they may have reasons to interact with the contract.

In our current design, we may need to implement two functions for `transfer`, one for encrypted call, and another for plain text call. I don't have good idea. What do you think?

kvinwang commented on standalone/pruntime/enclave/src/prpc_service.rs:
```diff
@@ -538,37 +555,126 @@ pub fn get_egress_messages(output_buf_len: usize) -> RpcResult<EgressMessages> {
+        }
+        info!("Verifying signature passed!");
+    }
+    let ecdh_key = LOCAL_STATE
+        .lock()
```
> In our current design, we may need to implement two functions for `transfer`, one for encrypted call, and another for plain text call. I don't have good idea. What do you think?

If this is a concret contract special issue, it's not a big problem to handle the same call on the two topics. If all contracts    have the requirements, we should change the design.

kvinwang replied:
> No need to change in this PR, but let's start to consider to consolidate all the states together

I consider to do some refactors after std ready.

kvinwang commented:
Some modifications done:
- Brought back the optinal secret message channel.
- Merged the event channel into the command channel.
- Removed the transaction status related codes.
- Use `bind_contract32!` for ContractCommands rather than `bind_topic!` to avoid some fake `impl BindTopic` types.
- Removed AccountIdWrapper.

@h4x3rotab pls review.

h4x3rotab commented on crates/phala-mq/src/types.rs:
```diff
@@ -53,8 +60,18 @@ impl MessageOrigin {
     }
+
+    /// Returns the account id if the origin is from a user, or `Err(BadOrigin)` otherwise
+    pub fn account(self) -> Result<AccountId32, BadOrigin> {
```
Suggest to rename it to `ensure_signed()` to match Substrate's convention

kvinwang replied:
> Suggest to rename it to `ensure_signed()` to match Substrate's convention

This is not on-chain code, So I think it's better to match Rust's convention.

h4x3rotab reviewed (approved):
LGTM
